File: web/src/lib/anomaly.ts
// "Is the grid cleaner or dirtier than usual right now?" — relative context from
// the accumulated history, which is more motivating than a bare absolute number.
//
// Pure and side-effect-free so it's unit-testable; the globe feeds it the history
// it already fetched, so there's no extra request.

export type UsualComparison = {
  status: "cleaner" | "typical" | "dirtier";
  deltaPct: number; // signed: negative = cleaner than usual
  basis: "hour" | "recent"; // compared against same-hour-of-day, or all recent points
  sampleSize: number;
};

type Point = { timestamp: string; carbon_intensity_gco2_kwh: number };

const MIN_HOUR_SAMPLES = 3;
const MIN_RECENT_SAMPLES = 6;
const DEADBAND_PCT = 10; // within ±10% reads as "typical"

function median(values: number[]): number {
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

// Compare `current` intensity to the historical baseline. Prefers the same
// hour-of-day (UTC) when there's enough of it (diurnal patterns dominate grid
// carbon); otherwise falls back to all recent points. Returns null when there
// isn't enough history yet to say anything honest.
export function relativeToUsual(
  current: number,
  points: Point[],
  now: Date,
): UsualComparison | null {
  const hour = now.getUTCHours();
  const sameHour = points
    .filter((p) => new Date(p.timestamp).getUTCHours() === hour)
    .map((p) => p.carbon_intensity_gco2_kwh);

  let sample: number[];
  let basis: "hour" | "recent";
  if (sameHour.length >= MIN_HOUR_SAMPLES) {
    sample = sameHour;
    basis = "hour";
  } else if (points.length >= MIN_RECENT_SAMPLES) {
    sample = points.map((p) => p.carbon_intensity_gco2_kwh);
    basis = "recent";
  } else {
    return null; // baseline still building
  }

  const baseline = median(sample);
  if (baseline <= 0) return null;

  const deltaPct = Math.round(((current - baseline) / baseline) * 100);
  const status =
    deltaPct <= -DEADBAND_PCT ? "cleaner" : deltaPct >= DEADBAND_PCT ? "dirtier" : "typical";
  return { status, deltaPct, basis, sampleSize: sample.length };
}

File: src/carbon_mesh/engine/anomaly.py
"""Relative-context analysis: is a zone cleaner or dirtier than usual right now?

Compares the current intensity to a baseline drawn from the published history
archive. Pure and side-effect-free so it's unit-testable; mirrors the frontend's
lib/anomaly.ts so the API and the UI agree.
"""

from __future__ import annotations

from datetime import datetime, timezone
from statistics import median

from carbon_mesh.engine.recurring import _parse_utc

_MIN_HOUR_SAMPLES = 3
_MIN_RECENT_SAMPLES = 6
_DEADBAND_PCT = 10.0


def _insufficient(sample_size: int) -> dict:
    return {
        "status": "insufficient_history",
        "basis": "insufficient",
        "baseline_gco2_kwh": None,
        "delta_pct": None,
        "sample_size": sample_size,
    }


def compute_anomaly(current: float, points: list[dict], now: datetime) -> dict:
    """Compare ``current`` to a baseline from history ``points`` (each ``{t,c,r}``).

    Prefers the same hour-of-day (UTC) when there's enough of it; otherwise all
    recent points; otherwise reports insufficient history. ``status`` is
    cleaner_than_usual / typical / dirtier_than_usual within a ±10% deadband.
    """
    target_hour = now.astimezone(timezone.utc).hour
    same_hour: list[float] = []
    all_values: list[float] = []
    for p in points:
        c = p.get("c")
        t = p.get("t")
        if c is None or not t:
            continue
        all_values.append(c)
        ts = _parse_utc(t)
        if ts is None:
            continue
        if ts.hour == target_hour:
            same_hour.append(c)

    if len(same_hour) >= _MIN_HOUR_SAMPLES:
        sample, basis = same_hour, "hour_of_day"
    elif len(all_values) >= _MIN_RECENT_SAMPLES:
        sample, basis = all_values, "recent"
    else:
        return _insufficient(len(all_values))

    baseline = median(sample)
    if baseline <= 0:
        return _insufficient(len(all_values))

    delta_pct = round((current - baseline) / baseline * 100, 1)
    if delta_pct <= -_DEADBAND_PCT:
        status = "cleaner_than_usual"
    elif delta_pct >= _DEADBAND_PCT:
        status = "dirtier_than_usual"
    else:
        status = "typical"
    return {
        "status": status,
        "basis": basis,
        "baseline_gco2_kwh": round(baseline, 1),
        "delta_pct": delta_pct,
        "sample_size": len(sample),
    }
